### PR TITLE
feat: GH Workflow Runs as Card View

### PR DESCRIPTION
**Pull request by blbose**
## Hey, I just made a Pull Request!

I contributed to have a separate page with different UI to show the workflow runs as card view with search, branch dropdown and paging option

Screenshot 2023-11-02 at 7 06 33 PM

#### :heavy_check_mark: Checklist

- [ ] [A changeset describing the change and affected packages. ([khaki-hounds-think.md](https://github.com/philips-forks/backstage/blob/feat-gh-actions-workflow-cards/.changeset/khaki-hounds-think.md)
- [ ] Added workflow-cards
- [ ] Tests for new functionality and regression tests for bug fixes - Test added
- [ ] Screenshots attached (for UI changes)
- [ ] All your commits have a `Signed-off-by` line in the message.

### CODE REVIEW

File: plugins/github-actions/src/components/Router.tsx [hunk 1]
```diff
@@ -41,7 +41,7 @@ export const Router = () => {
   }
   return (
     <Routes>
-      <Route path="/" element={<WorkflowRunsTable entity={entity} />} />
+      <Route path="/" element={<WorkflowRunsCard entity={entity} />} />
```
Comment (awanlin) [inline plugins/github-actions/src/components/Router.tsx:44]: From my reading this change would mean you would no longer be able to use the Table view. I'm not sure that everyone using this plugin would want that. Can you update this so both views are supported? That gives people the option to pick the one they want

File: .changeset/khaki-hounds-think.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+'@backstage/plugin-github-actions': minor
+---
+
+Github Workflow Runs UI is modified to show in Card view instead of table, with branch selection option
```
Comment (awanlin) [inline .changeset/khaki-hounds-think.md:5]: This should be updated to say it's an optional Card view, some may still want to use the Table view

File: plugins/github-actions/api-report.md [hunk 3]
```diff
@@ -142,6 +174,431 @@ export class GithubActionsClient implements GithubActionsApi {
     RestEndpointMethodTypes['actions']['getWorkflowRun']['response']['data']
   >;
   // (undocumented)
+  listBranches(options: {
+    hostname?: string;
+    owner: string;
+    repo: string;
+    page?: number;
+  }): Promise<
```
Comment (awanlin) [inline plugins/github-actions/api-report.md:182]: The return results seem very verbose and I'm not sure we are using all of this, can we trim this down to only what is needed?

File: plugins/github-actions/src/components/useWorkflowRuns.ts [hunk 4]
```diff
@@ -41,12 +42,12 @@ export function useWorkflowRuns({
   owner,
   repo,
   branch,
-  initialPageSize = 5,
+  initialPageSize = 6,
```
Comment (awanlin) [inline plugins/github-actions/src/components/useWorkflowRuns.ts:45]: More curious then anything but why is this change needed?

**Review by awanlin — CHANGES_REQUESTED** [review]
Thanks for the contribution @blbose, I like the idea seems like a good way to show this data but we need to make sure you can still use the Table view, I've mentioned this a few times in my review.

Also, there are several empty fles with changes that do not seem related. You'll want to clean that up. I might try rebasing on the latest master, that might help 👍

**Comment by blbose** [comment]
> Thanks for the contribution @blbose, I like the idea seems like a good way to show this data but we need to make sure you can still use the Table view, I've mentioned this a few times in my review.
> 
> Also, there are several empty fles with changes that do not seem related. You'll want to clean that up. I might try rebasing on the latest master, that might help 👍

Hi @awanlin, of course table view is much needed and the page which I created is a option to use in my organisation or anyone interested, they can call it in the router page.

Sure I will remove those files not sure how it got added.

File: plugins/github-actions/src/components/Router.tsx (hunk 1 above)
Comment (blbose) [reply plugins/github-actions/src/components/Router.tsx:44]: > From my reading this change would mean you would no longer be able to use the Table view. I'm not sure that everyone using this plugin would want that. Can you update this so both views are supported? That gives people the option to pick the one they want

Like mentioned above its only a option either table view or card view can be used here.., its our choice to use either of those, may be we can detailed the usage in ReadMe file, can you suggest??

File: .changeset/khaki-hounds-think.md (hunk 2 above)
Comment (blbose) [reply .changeset/khaki-hounds-think.md:5]: > This should be updated to say it's an optional Card view, some may still want to use the Table view

Agree, its optional card view

File: plugins/github-actions/src/components/useWorkflowRuns.ts (hunk 4 above)
Comment (blbose) [reply plugins/github-actions/src/components/useWorkflowRuns.ts:45]: Card view UI was looking good when it have 6 cards in one page than 5 to avoid empty space, so I changed to 6.

**Comment by blbose** [comment]
@awanlin, all changes done please review again and let me know if any changes required, regarding removing empty files added part of the commit, for some reason I'm not able to do it as the files in local or repo is not empty only in the commit its empty. 

Any help here is much appreciated.

File: plugins/github-actions/src/components/Router.tsx [hunk 5]
```diff
@@ -31,7 +32,8 @@ export const isGithubActionsAvailable = (entity: Entity) =>
   Boolean(entity.metadata.annotations?.[GITHUB_ACTIONS_ANNOTATION]);
 
 /** @public */
-export const Router = () => {
+// Optional cardview prop
+export const Router = ({ cardView = false }: { cardView?: boolean }) => {
```
Comment (Rugvip) [inline plugins/github-actions/src/components/Router.tsx:36]: Could we make this a string enum instead? Something like `view: 'table' | 'cards'`, and then default it to `'table'`. That allows for additional views in the future, and even expanding to view-specific options with `view: { type: 'view', ... } | ...`

File: packages/backend-plugin-manager/package.json
Comment (Rugvip) [inline packages/backend-plugin-manager/package.json:1]: Please revert these file mode changes

**Review by Rugvip — CHANGES_REQUESTED** [review]
Thank you! 👍 

Just a few nits

File: plugins/github-actions/README.md [hunk 6]
```diff
@@ -92,3 +92,35 @@ integrations:
 - There is a limit of 100 apps for one OAuth client/token pair
 - The OAuth application must be at the GitHub organization level in order to display the workflows. If you do
   not see any workflows, confirm the OAuth application was created in the organization and not a specific user account.
+
+## Optional Workflow Runs Card View
+
+Github WorkflowRuns optional UI to show in Card view instead of table, with branch selection option
+
+![Sonar Card](./docs/card-view.png)
```
Comment (awanlin) [inline plugins/github-actions/README.md:100]:
```suggestion
![Card view example](./docs/card-view.png)
```

**Review by awanlin — CHANGES_REQUESTED** [review]
Just took another look and found one small nit and then if you can follow up on @Rugvip's comments as well, please 👍

File: plugins/github-actions/README.md (hunk 6 above)
Comment (blbose) [reply plugins/github-actions/README.md]: Fixed

File: packages/backend-plugin-manager/package.json
Comment (blbose) [reply packages/backend-plugin-manager/package.json:1]: Reverted

File: plugins/github-actions/src/components/Router.tsx (hunk 5 above)
Comment (blbose) [reply plugins/github-actions/src/components/Router.tsx]: Changed to enum.

**Review by awanlin — APPROVED** [review]

**Comment by acierto** [comment]
There are few places in UI which could be fixed too

Screenshot 2023-11-11 at 12 47 54

Screenshot 2023-11-11 at 12 47 56

Icons are not really aligned with the text position

**Comment by acierto** [comment]
What does these 3 digits mean?

I can guess some of them, but from UX perspective making it crystal clear is more user friendly.

Screenshot 2023-11-11 at 12 49 10

User shouldn't guess or go to read the docs :)

**Comment by blbose** [comment]
> What does these 3 digits mean?
> 
> I can guess some of them, but from UX perspective making it crystal clear is more user friendly.
> 
> Screenshot 2023-11-11 at 12 49 10
> User shouldn't guess or go to read the docs :)

Fixed

**Comment by blbose** [comment]
> There are few places in UI which could be fixed too
> 
> Screenshot 2023-11-11 at 12 47 54 Screenshot 2023-11-11 at 12 47 56
> Icons are not really aligned with the text position

Fixed

**Comment by blbose** [comment]
> > There are few places in UI which could be fixed too
> > Screenshot 2023-11-11 at 12 47 54 Screenshot 2023-11-11 at 12 47 56
> > Icons are not really aligned with the text position
> 
> Fixed

File: plugins/github-actions/src/components/useWorkflowRuns.ts [hunk 7]
```diff
@@ -56,13 +57,51 @@ export function useWorkflowRuns({
...
   } = useAsyncRetry<WorkflowRun[]>(async () => {
+    const fetchedDefaultBranch = await api.getDefaultBranch({
+      hostname,
+      owner,
+      repo,
+    });
+    setDefaultBranch(fetchedDefaultBranch);
+    if (branch === 'default') {
+      // eslint-disable-next-line no-param-reassign
+      branch = fetchedDefaultBranch;
```
Comment (acierto) [inline plugins/github-actions/src/components/useWorkflowRuns.ts:77]: It's better to introduce a new variable then disabling the rule

Comment (blbose) [reply plugins/github-actions/src/components/useWorkflowRuns.ts]: Fixed.

File: plugins/github-actions/src/components/WorkflowRunsCard/WorkflowRunsCard.tsx [hunk 8]
```diff
@@ -0,0 +1,417 @@
...
+}: {
+  searchTerm: string;
+  handleSearch: (event: ChangeEvent<HTMLInputElement>) => void;
+  retry: (() => void) | (() => void) | (() => void) | (() => void);
+}) => {
+  return (
+    <>
+      <Box flexGrow={1} />
+      <TextField
+        id="search"
```
Comment (acierto) [inline plugins/github-actions/src/components/WorkflowRunsCard/WorkflowRunsCard.tsx:266]: It's better to avoid IDs if possible https://medium.com/@clairecodes/reasons-not-to-use-ids-in-css-a42204fb0d97#:~:text=Classes%20can%20be%20reused,of%20the%20advantages%20of%20CSS.

Comment (blbose) [reply plugins/github-actions/src/components/WorkflowRunsCard/WorkflowRunsCard.tsx:266]: @acierto, I'm using the IDs part of logic to get the search control and drop menu control, so it's needed for now, can we make this change part of next phase improvement?

Comment (Rugvip) [reply plugins/github-actions/src/components/WorkflowRunsCard/WorkflowRunsCard.tsx:266]: +1 to avoiding `id`. I'm not seeing where you end up using it btw? Typically you'd be able to use a ref instead.

Comment (blbose) [reply plugins/github-actions/src/components/WorkflowRunsCard/WorkflowRunsCard.tsx]: Fixed, my bad Id is not used anywhere.

**Comment by ScottGuymer** [comment]
@Rugvip is there anything else to add for this PR?

We are keen to get this merged and deployed soon in out instance.

File: .changeset/khaki-hounds-think.md [hunk 9]
```diff
@@ -0,0 +1,5 @@
+---
+'@backstage/plugin-github-actions': minor
```
Comment (Rugvip) [inline .changeset/khaki-hounds-think.md:2]:
```suggestion
'@backstage/plugin-github-actions': patch
```

Since this isn't a breaking change we can keep this a patch since the plugin is still on 0.x

File: plugins/github-actions/README.md [hunk 10]
```diff
@@ -92,3 +92,36 @@ integrations:
...
+
+2. Add to the app `EntityPage` component:
+
+```tsx
+// In packages/app/src/components/catalog/EntityPage.tsx
+import {
+  EntityGithubActionsContent,
+  isGithubActionsAvailable,
+  ViewEnum
+} from '@backstage/plugin-github-actions';
```
Comment (Rugvip) [inline plugins/github-actions/README.md:117]: These instructions are already listed above, I'd say keep this section more concise by just highlighting the additional prop

File: plugins/github-actions/src/components/Router.tsx [hunk 11]
```diff
@@ -23,30 +23,39 @@ import {
...
+import { RouterProps, ViewEnum } from '../api/types';
 
 /** @public */
 export const isGithubActionsAvailable = (entity: Entity) =>
   Boolean(entity.metadata.annotations?.[GITHUB_ACTIONS_ANNOTATION]);
 
 /** @public */
-export const Router = () => {
+export const Router = ({ view = ViewEnum.Table }: RouterProps) => {
   const { entity } = useEntity();
```
Comment (Rugvip) [inline plugins/github-actions/src/components/Router.tsx:37]:
```suggestion
export const Router = (props: RouterProps) => {
  const { view = ViewEnum.Table } = props;
  const { entity } = useEntity();
```

Avoid destructuring in the parameter list for public APIs, since that pollutes the declaration in docs

File: plugins/github-actions/src/api/types.ts [hunk 12]
```diff
@@ -49,3 +49,25 @@ export enum BuildStatus {
...
+
+/** @public */
+export enum ViewEnum {
+  Table = 'table',
+  Card = 'cards',
+}
+
+/** @public */
+export interface RouterProps {
+  view?: ViewEnum;
```
Comment (Rugvip) [inline plugins/github-actions/src/api/types.ts:72]:
```suggestion
  view?: 'cards' | 'table';
```

Let's avoid additional exports for the public API. It's fine to use an enum internally if you prefer, but let's avoid having the integrator import and extra symbol to configure this.

**Review by Rugvip — CHANGES_REQUESTED** [review]
Thank you, looks good! 👍 

Just a couple of nits on the presentation 😁

File: .changeset/khaki-hounds-think.md (hunk 9 above)
Comment (blbose) [reply .changeset/khaki-hounds-think.md]: Updated

File: plugins/github-actions/README.md (hunk 10 above)
Comment (blbose) [reply plugins/github-actions/README.md]: Updated and removed those instructions

File: plugins/github-actions/src/components/Router.tsx (hunk 11 above)
Comment (blbose) [reply plugins/github-actions/src/components/Router.tsx]: Fixed

**Comment by blbose** [comment]
@Rugvip, made all the requested changes.

**Comment by blbose** [comment]
@Rugvip, anything else to be added.

Can we go for a merge and deployment.

File: plugins/github-actions/src/components/WorkflowRunsCard/WorkflowRunsCard.tsx [hunk 13]
```diff
@@ -0,0 +1,416 @@
...
+                      severity={
+                        statusColors[
+                          run.conclusion as keyof typeof statusColors
+                        ] as Color
+                      }
+                      style={{ alignItems: 'center' }}
+                    >
+                      <Typography variant="h6">
+                        <Link
+                          component={RouterLink}
```
Comment (freben) [inline plugins/github-actions/src/components/WorkflowRunsCard/WorkflowRunsCard.tsx:158]: this really shouldn't be necessary (but also shouldn't hurt); Link already does this internally, dynamically, depending on the `to`.

File: plugins/github-actions/src/components/WorkflowRunsCard/WorkflowRunsCard.tsx [hunk 14]
```diff
@@ -0,0 +1,416 @@
...
+};
+
+const WorkflowRunsCardSearch = ({
+  searchTerm,
+  handleSearch,
+  retry,
+}: {
+  searchTerm: string;
+  handleSearch: (event: ChangeEvent<HTMLInputElement>) => void;
+  retry: (() => void) | (() => void) | (() => void) | (() => void);
```
Comment (freben) [inline plugins/github-actions/src/components/WorkflowRunsCard/WorkflowRunsCard.tsx:260]: Is... is this programmer humor or unintentional? 😅

File: plugins/github-actions/src/components/WorkflowRunsCard/WorkflowRunsCard.tsx [hunk 15]
```diff
@@ -0,0 +1,416 @@
...
+                    </ButtonGroup>
+                  </div>
+                </Box>
+              </Box>
+            </Box>
+          </Grid>
+        ))
+      ) : (
+        <Box p={2}>
+          {loading ? <CircularProgress /> : 'No matching runs found.'}
```
Comment (freben) [inline plugins/github-actions/src/components/WorkflowRunsCard/WorkflowRunsCard.tsx:228]: We have been pretty adamant about using `Progress` from `@backstage/core-components`, but ... in a card context like this, I guess this may be an ok fit

**Review by Rugvip — APPROVED** [review]
Thank you! 👍 

My comments are addressed, but I think it would be good to fix some of the nits from Freben

File: plugins/github-actions/src/components/WorkflowRunsCard/WorkflowRunsCard.tsx (hunk 14 above)
Comment (blbose) [reply plugins/github-actions/src/components/WorkflowRunsCard/WorkflowRunsCard.tsx]: unintentional :-) fixed.

**Comment by blbose** [comment]
> Thank you! 👍
> 
> My comments are addressed, but I think it would be good to fix some of the nits from Freben

@Rugvip, Freben comments are addressed.
